File: lab0/gemm.cpp
#include "gemm.h"
#include "methods/methods.h"

int main(){
  double *A = (double *)malloc(M * K * sizeof(double));
  double *B = (double *)malloc(K * N * sizeof(double));
  double *C = (double *)malloc(M * N * sizeof(double));

  // 初始化矩阵 A B C
  for (int i = 0; i < M * K; i++) A[i] = static_cast<double>(rand()) / RAND_MAX;
  for (int i = 0; i < K * N; i++) B[i] = static_cast<double>(rand()) / RAND_MAX;
  for (int i = 0; i < M * N; i++) C[i] = 0.0;

  chrono::time_point<chrono::high_resolution_clock> start_time = chrono::high_resolution_clock::now();

  #ifdef METHOD0
  method0(A, B, C);
  #endif
  #ifdef METHOD1
  method1(A, B, C);
  #endif
  #ifdef METHOD2
  method2(A, B, C);
  #endif
  #ifdef METHOD3
  method3(A, B, C);
  #endif

  chrono::time_point<chrono::high_resolution_clock> end_time = chrono::high_resolution_clock::now();

  chrono::duration<double> elapsed = end_time - start_time;

  double flops = 2.0 * N * N * N;
  double gflops = flops / elapsed.count();

  cout << "Time: " << elapsed.count() << " s" << endl << "GFLOPS: " << gflops << endl;

  return 0;
}

File: lab0/gemm.py
import random
import time

N = 2048

def randomMatrix(m, n):
  # 产生随机矩阵
  matrix = []
  for _ in range(m):
    row = []
    for _ in range(n):
      row.append(random.randint(0, 1000000) / 1000) # 浮点数
    matrix.append(row)
  return matrix

def naiveGEMM(A, B):
  # 矩阵乘法
  m = len(A)
  n = len(A[0])
  k = len(B[0])
  C = [[0 for _ in range(k)] for _ in range(m)]
  for i in range(m):
    for j in range(k):
      for l in range(n):
        C[i][j] += A[i][l] * B[l][j]
  return C

if __name__ == '__main__':
  m = N
  n = N
  k = N
  A = randomMatrix(m, n)
  B = randomMatrix(n, k)
  start_time = time.process_time()
  C = naiveGEMM(A, B)
  end_time = time.process_time()
  elapsed_time = end_time - start_time
  flops = 2 * N**3
  print("Time taken: ", elapsed_time)
  print("GFLOPS: ", (flops / elapsed_time) / 1e9)